crates/full-node/sov-rollup-apis/src/endpoints/dedup.rs

```diff
@@ -60,45 +64,125 @@ pub trait DeDupEndpoint<S: Spec>: Clone + Send + Sync + 'static {
     }
 }
 
-/// Provides the `/rollup/addresses/{address}/dedup` endpoint utilising the sovereign provided
+/// Provides the `/rollup/addresses/{address}/dedup` endpoint using the sovereign provided
 /// `uniqueness` module.
+///
+/// This endpoint supports two independent uniqueness mechanisms:
+/// - **Nonces**: Sequential counters that must be used in order (0, 1, 2, ...). No skipping allowed.
+/// - **Generations**: Non-sequential identifiers that can skip values (0, 3, 7, ...).
+///
+/// Both mechanisms are tracked independently per account, allowing flexible transaction ordering
+/// strategies. You can mix nonce and generation-based transactions for the same account.
 #[derive(Clone)]
-pub struct NonceDeDupEndpoint<S: Spec> {
+pub struct SovereignDeDupEndpoint<S: Spec> {
     state: ApiState<S>,
 }
 
-impl<S: Spec> NonceDeDupEndpoint<S> {
-    /// Creates a new `NonceDeDupEndpoint` instance.
+impl<S: Spec> SovereignDeDupEndpoint<S> {
+    /// Creates a new [`SovereignDeDupEndpoint`] instance.
     pub fn new(state: ApiState<S>) -> Self {
         Self { state }
     }
 }
 
-/// The response of the nonce module implementation.
-#[derive(serde::Serialize, Clone)]
-pub struct NonceResponse {
-    /// The current nonce assiociated with the requested address.
-    pub nonce: u64,
+/// The response of the Dedup implementation for both nonce and generation number.
+///
+/// Only one field will be populated based on the query parameter:
+/// - `nonce`: The next sequential nonce value (must be used in order)
+/// - `generation`: The next available generation number (can skip values)
+#[derive(Clone, serde::Serialize, serde::Deserialize)]
+pub struct DedupResponse {
+    /// The next nonce associated with the requested address.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub nonce: Option<u64>,
+    /// The next generation number associated with the requested address.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub generation: Option<u64>,
+}
+
+/// Query parameters for the dedup endpoint.
+#[derive(Debug, Clone, Default, serde::Deserialize)]
+pub struct DedupQuery {
+    /// Select which kind of uniqueness to return.
+    /// If omitted, returns Nonce
+    /// Example: `?select=nonce` or `?select=generation`
+    pub select: Option<SelectField>,
```

Comment:
Very minor nit but thoughts on naming the query parameter something like `type`? E.g. `?type=nonce` and `?type=generation`.
It will make the rust variable name annoying though so maybe not worth it

Reply:
Yeah, it is going to be nont nice and have to do serde attributes.